File: battlefield/handlers.py
from characters import Character

class SpecialAbilityHandler:
    def __init__(self, battlefield):
        self.battlefield = battlefield

    def execute(self, ability, caster:Character, target:Character):
        # Check MP before executing the ability.
        if caster.get_attribute("mp") < ability.mp_cost:
            self.battlefield.log(
                f"{caster.name} is out of chakra cannot activate {ability.name}.", dest="console")
            return None

        caster.modify_attribute("mp", -ability.mp_cost)
        target_name = target.name if target else "self"

        if ability.name == "evasion":
            return self._evasion(caster, ability.jutsu_name)
        elif ability.name == "critical strike":
            return self._critical_strike(caster, target, ability.jutsu_name)
        elif ability.name == "poison":
            return self._poison(caster, target, ability.jutsu_name)
        elif ability.name == "stun":
            return self._stun(caster, target, ability.jutsu_name)
        elif ability.name == "heal self":
            return self._heal_self(caster, ability.jutsu_name)
        elif ability.name == "heal others":
            return self._heal_others(caster, target, ability.jutsu_name)
        elif ability.name == "buff":
            return self._buff(caster, ability.jutsu_name)
        else:
            raise ValueError(
                f"Special ability '{ability.name}' is not implemented.")

    def _evasion(self, caster, jutsu_name):
        duration = 1  # active until the next attack
        self.battlefield.apply_status_effect(
            caster, "evasion", caster)
        self.battlefield.log(
            f"{caster.name} activates {jutsu_name} for {duration} turn(s), ", dest="console")
        return {"effect": "evasion",  "duration": duration}

    def _critical_strike(self, caster, target, jutsu_name):
        base_damage = caster.get_attribute("dmg")
        multiplier = caster.get_special_ability("critical strike").value
        extra_damage = base_damage * multiplier
        target.modify_attribute("hp", -extra_damage)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} on {target.name} dealing an extra {extra_damage} damage.", dest="console")
        return {"effect": "critical strike", "extra_damage": extra_damage}

    def _poison(self, caster, target, jutsu_name):
        poison_damage = 10
        duration = 3
        self.battlefield.apply_status_effect(
            caster, "poison", target)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} on {target.name} for {duration} turns (damage: {poison_damage}/turn).", dest="console")
        return {"effect": "poison", "damage": poison_damage, "duration": duration}

    def _stun(self, caster, target, jutsu_name):
        duration = 1
        self.battlefield.apply_status_effect(caster, "stun", target)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} {target.name}, causing them to miss their next turn.", dest="console")
        return {"effect": "stun", "duration": duration}

    def _heal_self(self, caster, jutsu_name):
        heal_amount = caster.get_special_ability("heal self").value
        caster.modify_attribute("hp", heal_amount)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} on self for {heal_amount} HP.", dest="console")
        return {"effect": "heal self", "heal_amount": heal_amount}

    def _heal_others(self, caster, target, jutsu_name):
        heal_amount = caster.get_special_ability("heal others").value
        caster_team = self.battlefield.get_team_by_player(caster)
        target = min(caster_team, key=lambda x:x.get_attribute("hp"))
        target.modify_attribute("hp", heal_amount)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} on {target.name} for {heal_amount} HP.", dest="console")
        return {"effect": "heal others", "heal_amount": heal_amount}

    def _buff(self, caster, jutsu_name):
        buff_amount = caster.get_special_ability("buff").value
        caster.modify_attribute("armor", buff_amount)
        self.battlefield.log(
            f"{caster.name} uses {jutsu_name} to buff their armor by {buff_amount}.", dest="console")
        return {"effect": "buff", "buff_amount": buff_amount}

File: src/classes/handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpecialAbilityHandler = void 0;
var SpecialAbilityHandler = /** @class */ (function () {
    function SpecialAbilityHandler(battlefield) {
        this.battlefield = battlefield;
    }
    SpecialAbilityHandler.prototype.execute = function (ability, caster, target) {
        var casterMP = caster.getAttribute('mp').getValue();
        var abilityName = ability.getName();
        if (casterMP < ability.getMPCost()) {
            this.battlefield.log("".concat(caster.getName(), " lacks MP to acivate ").concat(abilityName), 'console');
            return null;
        }
        caster.modifyAttribute('mp', casterMP - ability.getMPCost());
        switch (abilityName) {
            case 'evasion':
                return this.evasion(caster, abilityName);
            case 'critical strike':
                return this.criticalStrike(caster, target, abilityName);
            case 'poison':
                return this.poison(caster, target, abilityName);
            case 'stun':
                return this.stun(caster, target, abilityName);
            case 'heal self':
                return this.healSelf(caster, abilityName);
            case 'heal others':
                return this.healOthers(caster, abilityName);
            case 'buff':
                return this.buff(caster, abilityName);
            default:
                throw Error("".concat(abilityName, " is not implemented"));
        }
    };
    SpecialAbilityHandler.prototype.evasion = function (caster, abilityName) {
        var duration = 1;
        this.battlefield.applyStatusEffect(caster, 'evasion', caster);
        this.battlefield.log("".concat(caster.getName(), " activates ").concat(abilityName, " for  ").concat(duration, " turns"));
        return { effect: 'evasion', duration: duration };
    };
    SpecialAbilityHandler.prototype.criticalStrike = function (caster, target, abilityName) {
        var _a;
        var baseDamage = (_a = caster.getAttribute('dmg')) === null || _a === void 0 ? void 0 : _a.getValue();
        var multiplier = caster.getSpecialAbility('critical strike').getValue();
        if (!baseDamage)
            throw Error('dmg attribute does not exist');
        var extraDamage = baseDamage * multiplier;
        var targetHP = target.getAttribute('hp').getValue();
        target.modifyAttribute('hp', targetHP - extraDamage);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " on ").concat(target.getName(), " dealing an extra ").concat(extraDamage, " damage. ").concat(target.getName(), " has ").concat(targetHP - extraDamage, " HP left"), 'console');
        return { effect: 'critical strike', extraDamage: extraDamage };
    };
    SpecialAbilityHandler.prototype.poison = function (caster, target, abilityName) {
        var duration = 3;
        var poisonDamage = 10;
        this.battlefield.applyStatusEffect(caster, 'poison', target);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " on ").concat(target.getName(), " for  ").concat(duration, " turns (damage: ").concat(poisonDamage, "/turn)"), 'console');
        return { effect: 'poison', duration: duration, damage: poisonDamage };
    };
    SpecialAbilityHandler.prototype.stun = function (caster, target, abilityName) {
        var duration = 1;
        this.battlefield.applyStatusEffect(caster, 'stun', target);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " on ").concat(target.getName(), " causing them to miss their next turn"), 'console');
        return { effect: 'stun', duration: duration };
    };
    SpecialAbilityHandler.prototype.healSelf = function (caster, abilityName) {
        var healAmount = caster.getSpecialAbility('heal self').getValue();
        caster.modifyAttribute('hp', caster.getAttribute('hp').getValue() + healAmount);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " on self for ").concat(healAmount, " HP"), 'console');
        return { effect: 'heal self', healAmount: healAmount };
    };
    SpecialAbilityHandler.prototype.healOthers = function (caster, abilityName) {
        var healAmount = caster.getSpecialAbility('heal others').getValue();
        var teamMates = this.battlefield.getTeamByChar(caster).getMembers().filter(function (m) { return m !== caster; });
        var teamMatesAndHP = teamMates.map(function (teamMate) {
            return { teamMate: teamMate, hp: teamMate.getAttribute('hp').getValue() };
        });
        var target = teamMatesAndHP.sort(function (a, b) { return a.hp - b.hp; })[0];
        target.teamMate.modifyAttribute('hp', target.hp + healAmount);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " on ").concat(target.teamMate.getName(), " self for ").concat(healAmount, " HP"), 'console');
        return { effect: 'heal others', healAmount: healAmount };
    };
    SpecialAbilityHandler.prototype.buff = function (caster, abilityName) {
        var buffAmount = caster.getSpecialAbility('buff').getValue();
        caster.modifyAttribute('armor', caster.getAttribute('armor').getValue() + buffAmount);
        this.battlefield.log("".concat(caster.getName(), " uses ").concat(abilityName, " to buff their armor by ").concat(buffAmount), 'console');
        return { effect: 'buff', buffAmount: buffAmount };
    };
    return SpecialAbilityHandler;
}());
exports.SpecialAbilityHandler = SpecialAbilityHandler;
